okio-files/src/commonMain/kotlin/okio/Filesystem.kt

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio
+
+abstract class Filesystem {
```

Comment:
Is a filesytem strictly the view you would have outside of the process? e.g. a POSIX filesystem?  Is it assumed to be a singleton? Or is there some idea of virtual filesystems, mocking for testing, or federated views within an application?

Reply:
I’m thinking similar to OkHttp’s `Dns.SYSTEM`. There’s an obvious one, but if you wanna use DI, it’s very easy.

Reply:
I like this.  But means you aren't trying to deal with federation within this API. Java file systems are weird when jar ends up in a File, nothing works but eh abstraction of File is carrying data.

java.io.FileNotFoundException: file:/path/to/jarfile/bot.jar!/config/netclient.p (No such file or directory)

Reply:
No, federation is too difficult. Instead I think of `FileSystem` as being almost like a `Collection` type, like `Map`. You can interrogate and mutate state using `Path` as keys.